File: apps/api/src/services/companySnippets.client.ts
import { logger } from "../config/logger.js";

/**
 * Real public company-info lookup — no API key needed. Wikipedia's public
 * REST summary API (official, free, no scraping) plus the company's own
 * guessed homepage meta description. Deliberately does NOT scrape a search
 * engine's results page. Ported from HireConnect's search_client.py.
 */
const TIMEOUT_MS = 10_000;
const REQUEST_HEADERS = { "User-Agent": "Mozilla/5.0 (compatible; BelieveAI/1.0)" };
const META_DESCRIPTION_RE = /<meta[^>]+name=["']description["'][^>]+content=["']([^"']{20,300})["']/i;

function guessDomain(company: string): string | null {
  const cleaned = company.toLowerCase().replace(/[^a-z0-9]/g, "");
  return cleaned ? `${cleaned}.com` : null;
}

async function fetchWithTimeout(url: string, headers: Record<string, string>): Promise<Response> {
  const controller = new AbortController();
  const timer = setTimeout(() => controller.abort(), TIMEOUT_MS);
  try {
    return await fetch(url, { headers, signal: controller.signal });
  } finally {
    clearTimeout(timer);
  }
}

async function fetchWikipediaSummary(company: string): Promise<string | null> {
  try {
    const res = await fetchWithTimeout(
      `https://en.wikipedia.org/api/rest_v1/page/summary/${encodeURIComponent(company)}`,
      REQUEST_HEADERS,
    );
    if (res.status === 404) return null;
    if (!res.ok) return null;
    const data = (await res.json()) as { type?: string; extract?: string };
    // Disambiguation pages have no real "extract" worth trusting.
    if (data.type === "disambiguation") return null;
    return data.extract?.trim() || null;
  } catch (err) {
    logger.warn({ company, err }, "companySnippets: Wikipedia summary lookup failed");
    return null;
  }
}

async function fetchSiteMetaDescription(company: string): Promise<string | null> {
  const domain = guessDomain(company);
  if (!domain) return null;
  try {
    const res = await fetchWithTimeout(`https://${domain}`, REQUEST_HEADERS);
    if (!res.ok) return null;
    const html = await res.text();
    const match = META_DESCRIPTION_RE.exec(html);
    return match?.[1]?.trim() ?? null;
  } catch {
    return null;
  }
}

/** Best-effort — never throws. Missing snippets just means company intel synthesis
 * downstream marks everything low-confidence rather than fabricating values. */
export async function fetchCompanySnippets(company: string): Promise<string[]> {
  const [wiki, siteDescription] = await Promise.all([fetchWikipediaSummary(company), fetchSiteMetaDescription(company)]);
  return [wiki, siteDescription].filter((s): s is string => Boolean(s));
}

File: apps/backend/clients/company_snippets_client.py
"""Mirrors apps/api's companySnippets.client.ts — real public company-info
lookup, no API key needed. Wikipedia's public REST summary API (official,
free, no scraping) plus the company's own guessed homepage meta
description. Deliberately does NOT scrape a search engine's results page.
"""

from __future__ import annotations

import asyncio
import re
from urllib.parse import quote

import httpx

from core.logging import get_logger

logger = get_logger(__name__)

_TIMEOUT_SECONDS = 10.0
_REQUEST_HEADERS = {"User-Agent": "Mozilla/5.0 (compatible; BelieveAI/1.0)"}
_META_DESCRIPTION_RE = re.compile(r"""<meta[^>]+name=["']description["'][^>]+content=["']([^"']{20,300})["']""", re.IGNORECASE)


def _guess_domain(company: str) -> str | None:
    cleaned = re.sub(r"[^a-z0-9]", "", company.lower())
    return f"{cleaned}.com" if cleaned else None


async def _fetch_wikipedia_summary(company: str) -> str | None:
    try:
        async with httpx.AsyncClient(timeout=_TIMEOUT_SECONDS) as client:
            res = await client.get(f"https://en.wikipedia.org/api/rest_v1/page/summary/{quote(company, safe='')}", headers=_REQUEST_HEADERS)
        if res.status_code == 404 or res.status_code >= 400:
            return None
        data = res.json()
        # Disambiguation pages have no real "extract" worth trusting.
        if data.get("type") == "disambiguation":
            return None
        extract = (data.get("extract") or "").strip()
        return extract or None
    except Exception as err:  # noqa: BLE001 — best-effort lookup, caller treats a miss as "no snippet"
        logger.warning("companySnippets: Wikipedia summary lookup failed for %s: %s", company, err)
        return None


async def _fetch_site_meta_description(company: str) -> str | None:
    domain = _guess_domain(company)
    if not domain:
        return None
    try:
        async with httpx.AsyncClient(timeout=_TIMEOUT_SECONDS, follow_redirects=True) as client:
            res = await client.get(f"https://{domain}", headers=_REQUEST_HEADERS)
        if res.status_code >= 400:
            return None
        match = _META_DESCRIPTION_RE.search(res.text)
        return match.group(1).strip() if match else None
    except Exception:  # noqa: BLE001 — best-effort probe of a guessed domain
        return None


async def fetch_company_snippets(company: str) -> list[str]:
    """Best-effort — never raises. Missing snippets just means company intel
    synthesis downstream marks everything low-confidence rather than
    fabricating values."""
    wiki, site_description = await asyncio.gather(_fetch_wikipedia_summary(company), _fetch_site_meta_description(company))
    return [s for s in (wiki, site_description) if s]
